hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieDataBlock.java

```diff
@@ -224,7 +224,12 @@ public final <T> ClosableIterator<HoodieRecord<T>> getRecordIterator(List<String
       return allRecords;
     }
 
-    HashSet<String> keySet = new HashSet<>(keys);
+    // TODO(yihua): is this valid?
+
+    HashSet<String> keySet = new HashSet<>();
```

Comment:
If we can get back the records as a sorted iterator by key, then we can use the input keys iterator to filter in a more efficient way to comparing the head and popping instead of requiring to compute the hash of the key to check for presence in the set.